### PR TITLE
[stdlib][SR-7136] Fast-path ad-hoc printing of Strings

### PR DESCRIPTION
**Pull request by darquro**
I added dynamically checked whether something is a String and just print the string.

Resolves [SR-7136](https://bugs.swift.org/browse/SR-7136).

### CODE REVIEW

**Comment by darquro** [comment]
@swift-ci Please smoke test

File: stdlib/public/core/OutputStream.swift [hunk 1]
```diff
@@ -380,6 +380,11 @@ internal func _adHocPrint_unlocked<T, TargetStream : TextOutputStream>(
 internal func _print_unlocked<T, TargetStream : TextOutputStream>(
   _ value: T, _ target: inout TargetStream
 ) {
+  if let string = value as? String {
+    let mirror = Mirror(reflecting: string)
+    _adHocPrint_unlocked(string, mirror, &target, isDebugPrint: false)
```
Comment (jckarter) [inline stdlib/public/core/OutputStream.swift]: You can just do `target.write(string)`.

**Review by jckarter — CHANGES_REQUESTED** [review]

**Comment by jckarter** [comment]
Thanks for looking into this! _adHocPrint is probably not going to be much faster in practice, though. You can use `TextOutputStream`'s `write` method directly since we know we have a string here.

**Comment by darquro** [comment]
@jckarter Thank you for reviewing! I fixed it.

**Comment by jckarter** [comment]
Awesome, thanks! I'll run the tests and merge.

**Review by jckarter — APPROVED** [review]

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
Oh, `x as? String` will be true for `Optional<String>` as well, so this changes how Optional prints. I guess we'll have to move the `as? String` check after the `_isOptional` check below it.

**Comment by darquro** [comment]
OK, I'll fix it again. 👍

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Do we have anything to measure this speedup? I'm interested in detecting regressions and exploring further improvements here such as plumbing this higher up.

**Comment by milseman** [comment]
(I suspect the string builders are noisy)

@swift-ci please benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke test

**Comment by milseman** [comment]
@swift-ci please smoke test

**Comment by milseman** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke benchmark

**Comment by jckarter** [comment]
I suspect the primary benefit would be in startup time and memory usage, since this should make it so that `print("...")` doesn't warm up the conformance cache. I don't think our current benchmark suite measures that well, though.

**Comment by CodaFi** [comment]
Hmph.  Yeah, this looks like noise.

**Comment by milseman** [comment]
Do we have a benchmark (or should we write one) that will demonstrate the improvement?

**Comment by jckarter** [comment]
We might be able to rig something up that captures the warmup time aspect. We could definitely do with more memory usage benchmarks, but that might need more infrastructure.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by CodaFi** [comment]
I've filed https://bugs.swift.org/browse/SR-12572 about the benchmark infrastructure

@swift-ci test

**Comment by CodaFi** [comment]
⛵

**Pull request merged**